### PR TITLE
Emit a ``UserWarning`` and crash pylint if run in parallel and with custom plugin

### LINKED ISSUE #3232: fail/warn on using parallel execution with custom plugins
According to documentation:
http://pylint.pycqa.org/en/latest/user_guide/run.html#parallel-execution

> There are some limitations in running checks in parallel in the current implementation. It is not possible to use custom plugins (i.e. --load-plugins option)...

Actually, it is possible but silently broken.
`If this is still by design` then Pylint should inform a user about it in such cases.
As for now, I could run:
```
pylint -j 10 --load-plugins plugin_foo bar.py
```
without any warning or error.
Unfortunately, linting results are not the same as a single process linting, but Pylint silently pass. So, results are not predictable.

Proposal: emit a warning or better explicitly fail on using parallel execution with custom Pylint plugins, because people usually don't read the documentation while things works.

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #3232

As issue an as [pylint docs](state), users should not run pylint in parallel AND with custom plugins.

Looks like this now:
```
/Users/xxx/pylint/pylint/lint/run.py:189: UserWarning: Running pylint in parallel with custom plugins is not currently supported.
  warnings.warn(
```
and pylint then stops execution

### CODE REVIEW

**Comment by clavedeluna** [comment]
So there are legitimate test cases to update so they don't run both in parallel and with plugins, good to know... if a maintainer 👁️ this PR and thinks it's good, I'll update the tests.

**Comment by Pierre-Sassoulas** [comment]
Isn't the conclusion in the issue's discussion that we should crash ?

File: pylint/lint/run.py
Comment (clavedeluna) [inline pylint/lint/run.py:193]: Hm Ok I'm not super convinced anymore after realizing that if you run
`pylint test.py --enable=all --enable-all-extensions` 
on any test.py module raises this warning now...
which is prob because of `--enable-all-extensions`.  

Maybe the code needs to be that config.jobs can be either `0` or anything >1? that would make more sense...

**Comment by clavedeluna** [comment]
> Isn't the conclusion in the issue's discussion that we should crash ?

Yes, I did that in the code but didn't properly say that in the title + PR description, will update it.

```
 sys.exit(32)
```

**Review by cdce8p — CHANGES_REQUESTED** [review]
See https://github.com/PyCQA/pylint/issues/3232#issuecomment-1322790418

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
It seems 4 lines are not covered, would you mind adding a configuration functional tests, please ? (https://pylint.pycqa.org/en/latest/development_guide/contributor_guide/tests/writing_test.html#functional-tests-for-configurations)

File: pylint/lint/run.py
Comment (Pierre-Sassoulas) [inline pylint/lint/run.py:238]: We should move this condition up imo. If multiprocessing is not available then there's ni reason to crash because of it.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: pylint/lint/run.py
Comment (clavedeluna) [reply pylint/lint/run.py:238]: We don't crash if multiprocessing is None, see line 243...

Comment (Pierre-Sassoulas) [reply pylint/lint/run.py:238]: I meant if ``self.linter.config.jobs != 1`` and ``self._plugins`` and ``multiprocessing is None`` are all true, then we're exiting but we could keep linting because the multiprocessing is not going to be activated. We can deactivate the multiprocessing first instead.

```python
if multiprocessing is None:
    print(
        "Multiprocessing library is missing, fallback to single process",
        file=sys.stderr,
    )
    self.linter.set_option("jobs", 1)
if self.linter.config.jobs != 1:
            # Validate if jobs is 0 or positive num other than 1
            if self._plugins:
                warnings.warn(
                    "Running pylint in parallel with custom plugins is not currently supported.",
                    UserWarning,
                )
                sys.exit(32)
elif self.linter.config.jobs == 0:
                self.linter.config.jobs = _cpu_count()
```

Comment (clavedeluna) [reply pylint/lint/run.py:238]: I can sort of see the argument for your proposal but the point of checking `if multiprocessing is None:` inside `if self.linter.config.jobs != 1:` is to only print `print("Multiprocessing library is missing, fallback to single process",` when the user configured num of jobs is !=1. Moving the check up higher outside of the if statement, means we're going to print that regardless of the number of jobs, even if user passed `-j 1` which seems confusing to me...

anyway, once you read this if you still think it should be changed I will do so.

**Review by Pierre-Sassoulas — DISMISSED** [review]
👍

File: doc/whatsnew/fragments/3232.bugfix
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/3232.bugfix:1]:
```suggestion
Exit immediately with a ``UserWarning`` if pylint is run in parallel and with custom plugins. This can lead to inconsistent behaviors, and the multiprocessing is not very efficient due to most of the parsing our internal code representation generation not being parallelized anyway. Use ``--job=1`` instead.
```

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by clavedeluna** [comment]
This should be ready to merge

**Comment by Pierre-Sassoulas** [comment]
Let's wait for another maintainer review for this one.

**Comment by clavedeluna** [comment]
Closing since issue requirements have changed.

**Pull request closed without merging**
